### ISSUE TITLE
Jaxopt-equivalent L-BFGS

### ISSUE THREAD

**Issue by Mandelbr0t** [google-deepmind/optax#1201]
Hi All,

I am migrating L-BFGS from Jaxopt to Optax and struggle finding the equivalent of Jaxopts' L-BFGS in Optax.

How can I create the exact same L-BFGS optimizer in Optax given the following Jaxopt instantiation and arguments ?

```
import jaxopt
optim = jaxopt.LBFGS(
        fun=fun,
        has_aux=has_aux,
        verbose=optim_config.verbose,
        stepsize=stepsize,
        linesearch=linesearch,
        history_size=optim_config.history_size,
        maxiter=optim_config.maxiter,
        maxls=optim_config.maxls,
        tol=optim_config.tol,
    )
```

**Comment by rdyro** [comment]
Hi, thanks for the question, are you unable to create the optimizer or are you running into numerical discrepancy issues between implementations?

Have you seen the [intro to using optax](https://optax.readthedocs.io/en/latest/getting_started.html#basic-usage-of-optax) and [the L-BFGS doc including an example use](https://optax.readthedocs.io/en/latest/api/optimizers.html#optax.lbfgs)?

**Comment by Mandelbr0t** [comment]
Hi, thanks for the quick response! 
I have tried the example in your second link but got increasing loss at some stage during training, which does not happen in Jaxopt's L-BFGS (afaik it should not happen at all as long as the line search does its job).
Is it possible to brute-force the answer to my question using optax's alias_test.py tests against scipy.minimize? 
I remember that there is a static difference between jaxopt.lbfgs and scipy.minimize.lbfgs.

**Comment by SNMS95** [comment]
@Mandelbr0t Did you manage to fix the issue with non-decreasing loss?
I am having the issue too.
I will try to create a MWE but I am using LBFGS with meta-learning, thus, there is a lot of boilerplate code that I have to undo.

**Comment by vroulet** [comment]
@Mandelbr0t, @SNMS95,
If the objective is increasing, it means that the maximal number of linesearch steps is too small a priori.
Consider also using the verbose option in the linesearch and post here the result. 
A MWE would be greatly appreciated to pinpoint your issue.

And to answer maybe your first question: the jaxopt and optax implementations differ slightly as the current optax implementation takes a guess of stepsize of one at every iteration and uses a clipped gradient at the very first iteration.

**Comment by vroulet** [comment]
@Mandelbr0t, @SNMS95 any updates?
Bugs in solvers are worrisome so any reproducing example would be greatly welcome.

**Comment by SNMS95** [comment]
@vroulet  For me, it was the failure of line-search that caused the spikes!
But I found that I cannot use line-search as the inner optimizer when doing meta-learning since It is implemented  with
`jax.lax.while_loop`, preventing reverse-mode differentiation.
Is there any way to differentiate through line-search ?

```
from collections.abc import Callable
from typing import Iterator, Tuple
import chex
import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np
import optax

# MAke a data generator
def generator() -> Iterator[Tuple[chex.Array, chex.Array]]:
  rng = jax.random.PRNGKey(0)

  while True:
    rng, k1, k2 = jax.random.split(rng, num=3)
    x = jax.random.uniform(k1, minval=0.0, maxval=10.0)
    y = 10.0 * x + jax.random.normal(k2)
    yield x, y

g =  generator()

# "NN"
def f(theta: chex.Array, x: chex.Array) -> chex.Array:
  return x * theta

theta = jax.random.normal(jax.random.PRNGKey(42))

# Setup optimizers
init_learning_rate = jnp.array(0.1)
meta_learning_rate = jnp.array(0.03)

opt_prev = optax.inject_hyperparams(optax.sgd)(learning_rate=init_learning_rate)
opt = optax.chain(opt_prev, 
                  optax.scale_by_zoom_linesearch(max_linesearch_steps=15)
                  )
meta_opt = optax.adam(learning_rate=meta_learning_rate)

# Loss functions
def loss(theta, x, y):
  return optax.l2_loss(y, f(theta, x))

def step(theta, state, x, y):
  val, grad = jax.value_and_grad(loss)(theta, x, y)
  updates, state = opt.update(grad, state, theta,
                              grad=grad, value=val,
                              value_fn = lambda a: loss(a, x, y))
  theta = optax.apply_updates(theta, updates)
  return theta, state

def outer_loss(eta, theta, state, samples):
  state[0].hyperparams['learning_rate'] = jax.nn.sigmoid(eta)

  for x, y in samples[:-1]:
    theta, state = step(theta, state, x, y)

  x, y = samples[-1]
  return loss(theta, x, y), (theta, state)

@jax.jit
def outer_step(eta, theta, meta_state, state, samples):
  grad, (theta, state) = jax.grad(
      outer_loss, has_aux=True)(eta, theta, state, samples)
  meta_updates, meta_state = meta_opt.update(grad, meta_state)
  eta = optax.apply_updates(eta, meta_updates)
  return eta, theta, meta_state, state

state = opt.init(theta)
# inverse sigmoid, to match the value we initialized the inner optimizer with.
eta = -np.log(1. / init_learning_rate - 1)
meta_state = meta_opt.init(eta)
N = 7
learning_rates = []
thetas = []

for i in range(50):
  samples = [next(g) for i in range(N)]
  eta, theta, meta_state, state = outer_step(eta, theta, meta_state, state, samples)
  learning_rates.append(jax.nn.sigmoid(eta))
  thetas.append(theta)

plt.plot(learning_rates)

```

**Comment by SNMS95** [comment]
@vroulet 

```
import optax
import jax
import jax.numpy as jnp

def fun(x):
    return jnp.sum(x**2)

x = jnp.array([1.0, 2.0, 3.0])
# optimizer
opt = optax.chain(optax.clip_by_global_norm(0.0001), optax.lbfgs())
opt_state = opt.init(x)
losses = []
for i in range(20):
    value, grad = jax.value_and_grad(fun)(x)
    updates, opt_state = opt.update(grad, opt_state, x, grad=grad, value=value,
                                    value_fn=fun)
    x = optax.apply_updates(x, updates)
    losses.append(value)

print(x)
print(fun(x))
import matplotlib.pyplot as plt
plt.plot(losses)
```
This causes spikes as well but I am clipping the gradient before, which might be a reason since `value_fn` is not clipped.

**Comment by vroulet** [comment]
> Is there any way to differentiate through line-search ?

You may use the implicit function theorem and code it with a custom jvp, see jaxopt for example. Optimistix has support for differentiable while loops too. In general, LBFGS is not coded to be backpropagated through indeed.

> This causes spikes as well but I am clipping the gradient before, which might be a reason since value_fn is not clipped.

If the gradients you're feeding do not match the value of the function you are minimizing, the solver has no reason to work properly. Typically, by clipping the gradients you may end up having a direction for optimization that is not a descent direction so the linesearch simply cannot work. The "verbose" option in the linesearch could probably help you see that.

Out of curiosity, did you take a look at the [notebook](https://optax.readthedocs.io/en/latest/_collections/examples/lbfgs.html) explaining briefly lbfgs?

**Comment by vroulet** [comment]
Ok I'll close this stale issue. Feel free to reopen if necessary!
